Sources/ComposableArchitecture/Reducer/Reducers/Scope.swift

```diff
@@ -0,0 +1,278 @@
+/// it can update the parent state and execute effects.
+///
+/// For example, given the basic scaffolding of child reducer:
+///
+/// ```swift
+/// struct Child: ReducerProtocol {
+///   struct State {
+///     // ...
+///   }
+///   enum Action {
+///     // ...
+///   }
+///   // ...
+/// }
+/// ```
+///
+/// A parent reducer with a domain that holds onto the child domain can use
+/// ``init(state:action:_:)`` to embed the child reducer in its
+/// ``ReducerProtocol/body-swift.property-7foai``:
+///
+/// ```swift
+/// struct Parent: ReducerProtocol {
+///   struct State {
+///     var child: Child.State
+///     // ...
+///   }
+///
+///   enum Action {
+///     case child(Child.Action)
+///     // ...
+///   }
+///
+///   var body: some ReducerProtocol<State, Action> {
+///     Scope(state: \.child, action: /Action.child) {
+///       Child()
+///     }
+///     Reduce { state, action in
+///       // Additional parent logic and behavior
+///     }
+///   }
+/// }
+/// ```
+///
+/// ## Enum state
+///
+/// The ``Scope`` reducer also works when state is modeled as an enum, not just a struct. In that
+/// case you can use ``init(state:action:_:file:fileID:line:)`` to specify a case path that
+/// identifies the case of state you want to scope to.
+///
+/// For example, if your state was modeled as an enum for unloaded/loading/loaded, you could
+/// scope to the loaded case to run a reduce on only that case:
+///
+/// ```swift
+/// struct Feature: ReducerProtocol {
+///   enum State {
+///     case unloaded
+///     case loading
+///     case loaded(Child.State)
+///   }
+///   enum Action {
+///     case child(Child.Action)
+///     // ...
+///   }
+///
+///   var body: some ReducerProtocol<State, Action> {
+///     Scope(state: /State.loaded, action: /Action.child) {
+///       Child()
+///     }
+///     Reduce { state, action in
+///       // Additional feature logic and behavior
+///     }
+///   }
+/// }
+/// ```
+///
+/// It is important to note that the order of combine ``Scope`` and your additional feature logic
+/// matters. It must be combined before the additional logic. In the other order it would be
+/// possible for the feature to intercept a child action, switch the state to another case, and
+/// then the scoped child reducer would not be able to react to that action. That can cause subtle
+/// bugs, and so we show a runtime warning in that case, and cause test failures.
+///
+/// For an alternative to using ``Scope`` with state case paths that enforces the order, check out
+/// the ``ifCaseLet(_:action:then:file:fileID:line:)`` operator.
+public struct Scope<ParentState, ParentAction, Child: ReducerProtocol>: ReducerProtocol {
+  public enum StatePath {
+    case casePath(
+      CasePath<ParentState, Child.State>,
+      file: StaticString,
+      fileID: StaticString,
+      line: UInt
+    )
+    case keyPath(WritableKeyPath<ParentState, Child.State>)
+  }
+
+  public let toChildState: StatePath
```

Comment:
Does this kind of property need to be `public`?
As an aside, I've just discovered the [`@_documentation(visibility: internal)`](https://github.com/apple/swift/pull/60242) attribute which doesn't seem to work yet in Xcode 14.1b3.